Start to add MXH treatment for GS2

d7919 opened the pull request:
Attempting to add support for MXH with GS2. @ZedThree added this into GS2 itself since version 8.2.0 and is perhaps better placed to address what the pyro changes should actually look like but I thought I'd get the ball rolling.

daviddickinson committed:
Start to add MXH treatment for GS2

d7919 commented:
I'm also trying to tweak the mxh inputs for the next release of GS2 through https://bitbucket.org/gyrokinetics/gs2/pull-requests/1075 so we might want to just introduce support from GS2 8.3?

bpatel2107 commented:
Nice! I'm guessing this will pass if you merge unstable back into this having merge #401

bpatel2107 commented on src/pyrokinetics/gk_code/gs2.py:
```diff
@@ -67,6 +82,15 @@ class GKInputGS2(GKInput, FileReader, file_type="GS2", reads=GKInput):
 
+    pyro_gs2_mxh_defaults = {
+        **pyro_gs2_miller_defaults,
+        "c0": 0.0,
```
Why is there a `c0`? Is that not the first value in `cn`?

daviddickinson committed:
Merge remote-tracking branch 'origin/unstable' into feature/add_mxh_for_gs2

bpatel2107 commented on src/pyrokinetics/gk_code/gs2.py:
```diff
@@ -688,11 +760,20 @@ def set(
-        # Assign Miller values to input file
-        for key, val in self.pyro_gs2_miller.items():
-            self.data[val[0]][val[1]] = local_geometry[key]
+            # Do we need to deal with rho factor in derivatives here?
```
Maybe? How are they defined in GS2, in pyro they are just $\frac{\partial c_n}{\partial \rho}$. No factor of $\rho$

d7919 committed:
Merge branch 'unstable' into feature/add_mxh_for_gs2

d7919 committed:
[skip ci] Apply black changes

d7919 commented on src/pyrokinetics/gk_code/gs2.py:
```diff
@@ -62,6 +64,15 @@ class GKInputGS2(GKInput, FileReader, file_type="GS2", reads=GKInput):
 
+    pyro_gs2_mxh = {
+        **pyro_gs2_miller,
+        "c0": ["theta_grid_parameters", "c0_mxh"],
```
GS2 no longer has a separate c0 input so we should remove this and similar later variables.

d7919 commented on src/pyrokinetics/gk_code/gs2.py:
```diff
@@ -251,6 +273,52 @@ def get_local_geometry_miller(self) -> LocalGeometryMiller:
+        if self.data["theta_grid_eik_knobs"]["irho"] != 2:
+            raise RuntimeError(
+                "Pyrokinetics requires GS2 input files to use "
+                "theta_grid_eik_knobs.bishop = 2"
```
Typo in message `bishop` --> `irho`

d7919 committed:
Update gs2.py to remove c0_mxh and add n_mxh

d7919 committed:
Update gs2.py to fix typo

d7919 committed:
Update gs2.py to fix error message typo

d7919 commented on src/pyrokinetics/gk_code/gs2.py:
```diff
@@ -72,6 +83,15 @@ class GKInputGS2(GKInput, FileReader, file_type="GS2", reads=GKInput):
+        "sn": [0.0, 0.0, 0.0, 0.0, 0.0],
+        "dcndr": [0.0, 0.0, 0.0, 0.0, 0.0],
+        "dsndr": [0.0, 0.0, 0.0, 0.0, 0.0],
+        "n_mxh": 5,
```
I don't know if we actually want to have the number of modes in this data set or if we should automatically determine it from the size of the cn/sn etc. lists.

bpatel2107 replied:
At the moment the defaults aren't massively important as they will likely get overwritten. As long as we set `n_moments` appropriately when creating the `LocalGeometryMXH` object it should be fine.

d7919 commented:
I think this might be ready for review now, although I've not tested this in any way

bpatel2107 commented on src/pyrokinetics/gk_code/gs2.py:
```diff
@@ -251,6 +273,52 @@ def get_local_geometry_miller(self) -> LocalGeometryMiller:
+        ):
+            mxh_data[pyro_key] = self.data[gs2_param].get(gs2_key, gs2_default)
+
+        # Do we need to scale derivatives by rho?
```
The definition in pyro is 

`pyro.local_geometry.cn =` $\partial c_n/\partial r$

If the GS2 input is actually $r \partial c_n/\partial r$ then it needs rescaling.

bpatel2107 commented:
This all looks mostly good. We have some roundtrip testing here

https://github.com/pyro-kinetics/pyrokinetics/blob/unstable/tests/test_roundtrip.py#L439

We could add GS2 to here. We would also need to tweak this file to allow for GS2 input file when using MXH

https://github.com/pyro-kinetics/pyrokinetics/blob/unstable/docs/examples/example_JETTO.py

@d7919 Would you be happy to add those in?

d7919 committed:
Update example_JETTO.py to allow GS2 input for non-Miller

Removes restriction on writing the GS2 input file when the geometry is not Miller

d7919 committed:
Update test_roundtrip.py to add GS2 to MXH roundtrip

d7919 commented:
@bpatel2107 I should hopefully have added what you've requested now.

d7919 committed:
Update gs2.py to allow LocalGeometryMXH

d7919 committed:
[skip ci] Apply black changes

d7919 committed:
Update gs2.py to use correct name for the number of moments

d7919 committed:
Update gs2.py to fix typo

d7919 committed:
Update gs2.py to define delta, s_kappa and s_delta in mxh

d7919 committed:
[skip ci] Apply black changes

d7919 committed:
Update gs2.py to add dZ0dr

d7919 committed:
Update gs2.py to convert mxh lists to arrays

d7919 committed:
Update gs2.py to fix name of instance referenced

d7919 committed:
Update gs2.py to be less general about how we convert list to array

d7919 committed:
[skip ci] Apply black changes

bpatel2107 reviewed (approved):
LGTm, was a long road though
